src/query/algorithms/entityresolution/EntityResolver.cpp

```diff
@@ -0,0 +1,502 @@
+        }
+    //Convert bloom filter to appropriate string
+        string filterStr = attrFilter.m_bits.to_string();
+        cout << filterStr << endl;
+        filterStr = utils.replace(filterStr, "0", ",0");
+        filterStr = utils.replace(filterStr, "1", ",1");
+        attrFilters[entity.first] = filterStr;
+
+    //Create structural filter
+        BloomFilter structFilter(256, 4);
+    //For each neighbour add selected attribute to bloom filter
+        for (auto neighbour: neighborhoodData[entity.first]) {
+            string selectedAttr = entityData[neighbour][0];
+            structFilter.insert(selectedAttr);
+        }
+    //Convert bloom filter to appropriate string
+        filterStr = structFilter.m_bits.to_string();
+        filterStr = utils.replace(filterStr, "0", ",0");
+        filterStr = utils.replace(filterStr, "1", ",1");
+        structFilters[entity.first] = filterStr;
+    }
+
+    return {attrFilters, structFilters};
+}
+
+/**
+ * Cluster created and collected Bloom filters of a given graph ID into a desired number of clusters
+ * @param graphID Graph ID of corresponding graph
+ * @param partitionCount The count of the (collected) partitions for the graph
+ * @param noClusters Desired clustering amount
+ */
+void EntityResolver::clusterFilters(string graphID, int partitionCount, int noClusters) {
+
+    Utils utils;
+    int noWorkers = stoi(utils.getJasmineGraphProperty("org.jasminegraph.server.nworkers"));
+    string partitionID = "0"; // use first worker related partitons
+    std::string attrFilterPath = "";
+    //std::string structFilterPath = "";
+
+    attrFilterPath =
+            utils.getJasmineGraphProperty("org.jasminegraph.server.instance.entityresolutionfolder") + "AttrFilters_" +
+            graphID + "_" + partitionID + ".txt";
+    //structFilterPath = utils.getJasmineGraphProperty("org.jasminegraph.entity.location")+"structfilters_" + graphID + "_" + partitionID + ".txt";
+
+    Mat<float> data;
+    data.load(attrFilterPath);
+
+    for (int i = 1; i < partitionCount; i++) {
+
+        string filepath = "attrfilters_" + graphID + "_" + to_string(i) + ".txt";
+        Mat<float> workerData;
+        workerData.load(filepath);
+        data = join_cols(data, workerData);
+    }
+
+    Mat<float> ids = data.col(0);
+    data.shed_col(0);
+    inplace_trans(data, "lowmem");
+
+    Kmeans<float> model(noClusters);
+    model.fit(data, 10);
+
+    //Apply clustering to bloom filters
+    Mat<short> pred = model.apply(data);
+
+    //Re-transpose data for saving
+    inplace_trans(data, "lowmem");
+    //Join with graph ids again
+    data = join_rows(ids, data);
+    //For each cluster write bloom filters of said cluster into a separate file
+    int clusterCount = model.getMeans().n_cols;
+    //Separate attr filters into clusters
+    seperateClusters(data, pred, clusterCount, "AttrCluster_" + graphID);
+    data.clear();
+    //Separate struct filters into clusters
```

Comment:
This part of the code seems not completed yet. Why do we have the commented lines?

Reply:
Commented lines removed

Reply:
Ok